movement/utils/vector.py

```diff
@@ -165,9 +166,182 @@ def pol2cart(data: xr.DataArray) -> xr.DataArray:
     ).transpose(*dims)
 
 
+def compute_signed_angle_2d(
+    test_vector: xr.DataArray, reference_vector: xr.DataArray | npt.NDArray
+) -> xr.DataArray:
+    r"""Compute the signed angle between two 2D vectors.
+
+    Angles are returned in radians, spanning :math:`(-\pi, \pi]` according to
+    the arctan2 convention. The sign of the angle follows the sign of the 2D
+    cross product ``test_vector``  :math:`\times` ``reference_vector``
+    and depends on the orientation of the coordinate system.
+
+    Parameters
+    ----------
+    test_vector : xarray.DataArray
+        An array of position vectors containing the ``space``
+        dimension with only ``"x"`` and ``"y"`` coordinates.
+    reference_vector : xarray.DataArray | numpy.ndarray
+        A 2D vector (or array of 2D vectors) against which to
+        compare ``test_vector``. May either be an xarray
+        DataArray containing the ``space``  dimension or a numpy
+        array containing one or more 2D vectors. (See Notes)
+
+    Returns
+    -------
+    xarray.DataArray :
+        An xarray DataArray containing signed angle between
+        ``test_vector`` and ``reference_vector`` for every
+        time-point. Matches the dimensions of ``test_vector``,
+        but without the ``space`` dimension.
+
+    Notes
+    -----
+    If passed as an xarray DataArray, the reference vector must
+    have the spatial coordinates ``x`` and ``y`` only, and must
+    have a ``time`` dimension matching that of the test vector.
+
+    If passed as a numpy array, the reference vector must have
+    one of three shapes:
+
+    - ``(2,)``: where dimension ``0`` contains spatial
+      coordinates (x,y), and no time dimension is specified.
+    - ``(1,2)``:, where dimension ``0`` corresponds to a
+      single time-point and dimension ``1`` contains spatial
+      coordinates (x,y).
+    - ``(n,2)``: where dimension ``0`` corresponds to
+      time and dimension ``1`` contains spatial coordinates
+      (x,y), and where ``n == len(test_vector.time)``.
+
+    Reference vectors containing more dimensions, or with shapes
+    otherwise different from those defined above are considered
+    invalid.
+
+    """
+    # test_vector checks
+    validate_dims_coords(test_vector, {"space": ["x", "y"]}, exact_coords=True)
+
+    # reference_vector checks
+    reference_vector_xr = _validate_reference_vector(
+        reference_vector, test_vector
+    )
+
+    test_unit = convert_to_unit(test_vector)
+    test_x = test_unit.sel(space="x")
+    test_y = test_unit.sel(space="y")
+
+    ref_unit = convert_to_unit(reference_vector_xr)
+    ref_x = ref_unit.sel(space="x")
+    ref_y = ref_unit.sel(space="y")
+
+    signed_angles = np.arctan2(
+        test_y * ref_x - test_x * ref_y,  # cross product test x ref
```

Comment:
```suggestion
        test_y * ref_x - test_x * ref_y,  # cross product ref x test
```

Little bit confused here; but "which way" is this angle computed? Are we computing the signed angle from `test_vector` to `reference_vector`, or `reference_vector` to `test_vector`?

My naive mathematical understanding here is that the signed angle between two vectors $a$ and $b$, IE the rotation that needs to be applied to $a$ to obtain $b$, is given by

$$ \mathrm{arctan2}( a\times b, a\cdot b ) = \mathrm{arctan2}( a_x b_y - a_y b_x, a_x b_x + a_y b_y )
$$

The function seems to be finding the signed angle that when applied to the `reference_vector` would obtain the `test_vector`. However this comment seems to be implying the opposite? This doesn't seem to be documented in the docstring (or if it is, maybe it should be made a bit clearer), so I wanted to check that this is what we're expecting to happen?

Of course, I may be misunderstanding :sweat_smile:

Reply:
I was also confused by this bit to be honest and unsure about which way we are actually doing vs which way we want to do it.

I agree with what you called "naive mathematical intuition", and the formula you gave, so I'd say we should go with that, and make sure implementation, comments and docstrings all agree with each other.

We could even include that formula in the docstring "Notes" (since Sphinx supports latex via the `math` directive).

Reply:
Perhaps we should even forgo the `test_vector` / `reference_vector` terminology here, to avoid confusion, and just use `vector_a` / `vector_b`, or even `a`/`b`.

Reply:
That said we should keep `reference_vector` in the higher level function in `kinematics`.

Reply:
Will do - have changed `compute_signed_angle_2d` to a purely mathematical implementation. 

This does mean we need to decide on a convention for computing the "heading angle" though. Is the "heading angle" the angle that when the forward-direction is rotated by it, gives the reference direction, or should it be the opposite (the angle that when the reference direction is rotated by it, gives the forward direction)? I can just add a toggle for this as an optional argument, but we should probably figure out which one is the default in the scientific literature. If there's a standard for this already then let me know and I'll go with that!

Reply:
I'm not aware of there being a standard for this, I don't know if @sfmig knows otherwise.

In my brain it's the angle that the "reference vector" is rotated by to match the "test vector".

So if you tell me the head angle is 45 degrees, in my mental model I will start from the reference direction and perform a 45-degree positive rotation (following the right-hand rule) to land on the head angle. But my view may be biased.

Tbh, I'm not sure a toggle is worth it, we can just pick one option and document it. We can add the toggle later if people start asking for the opposite direction.

Reply:
> In my brain it's the angle that the "reference vector" is rotated by to match the "test vector"

Yeah this was my gut too. A toggle isn't that much extra complexity though, and it might even be helpful for users (and us!) to see such an option is available. EG:

```python
def compute_heading_angle(
    data: xr.DataArray,
    left_keypoint: str,
    right_keypoint: str,
    reference_vector: npt.NDArray | list | tuple = (1, 0),
    camera_view: Literal["top_down", "bottom_up"] = "top_down",
    in_radians=False,
    angle_rotates: Literal["ref to head", "head to ref"] = "ref to head" # then switch the default when we know which is more common
) -> xr.DataArray:

Reply:
> Calling it compute_forward_angle(), doesn't sound very helpful though.

In order of increasing verbosity:

- `forward_vector_angle`
- `forward_vector_angle_relative_to`
- `perpendicular_heading_angle` (since this is abstractly between a `left` and `right` keypoint

Reply:
I'm fine with the function signature as you've outlined it.

Regarding the function name, I don't particularly like any of the options, but `compute_forward_vector_angle` is probably the most consistent with our current terminology (in other functions), as in we compute which direction is forward (for the animal), and we take the angle of that direction (with respect to some reference). I don't think it's worth saying "relative_to" because all angles are relative to sth.

Since in most cases people will be using this function to compute either the head or the body angle, I wonder whether it's worth adding two "aliases" —`compute_head_angle` and `compute_body_angle`—that will simply call `compute_forward_vector_angle`. The only difference would be in the docstring, which would either reference the head or the body explicitly (for the left and right keypoints). This is consistent with how we've treated [`compute_head_direction_vector()`](https://github.com/neuroinformatics-unit/movement/blob/15b3f41f3afc0673c3bc54686b3ec27e4fc37f89/movement/kinematics.py#L310), which is simply an alias for `compute_forward_vector()`.

I realise that such aliases are wholly unnecessary from a technical point of view, but may aid the findability of features and the readability of analysis workflow written with `movement`. Note that our collaborators have specifically asked for these two angles.

Feel free to push back on the aliases, or to leave them for a later PR, as they are kind of a scope-creep.

Reply:
> Feel free to push back on the aliases, or to leave them for a later PR, as they are kind of a scope-creep.

I'll open an issue once this is merged to add in the aliases - you're right that we should probably alias the most common functions, but I also agree that the focus of this is just to get the technical functionality in.

Reply:
Agreed.